### ISSUE TITLE
alignify doesn't need dollars, doesn't align right with multiple =, doesn't recognise <equality operators>, doesn't stack

### ISSUE THREAD

**Issue by polyluxus** [GaurangTandon/MarkdownShortcutsForSE#5]
## `align` doesn't need dollars  
If MathJax is set up to parse environments [`processEnvironments: true`](http://docs.mathjax.org/en/latest/options/preprocessors/tex2jax.html#configure-tex2jax), like it is on Chemistry.se, then adding `$$ ... $$` around the align just creates more noise than necessary.

## doesn't align right with multiple `=` in the same line
Consider the following scenario:

```
\psi = a_1\phi_1 + a_2\phi, where a_1^2 + a_2^2 = \sqrt2
```
with alignify there is one ampersand missing:
```
$$\begin{align}
\psi &= a_1\phi_1 + a_2\phi, where a_1^2 + a_2^2 &= \sqrt2
\end{align}$$
```
Even though I would not have a single clue how to prevent this behaviour/ solve this problem. (In this case the `&` needs to be inserted after the `,`, but that is not always the case. 

## doesn't recognise equality
Consider
```
a = b
a > b
a \geq b
```
which should recognise (the most common) relational operators to transform to
```
\begin{align}
a &= b\\
a &> b\\
a &\geq b
\end{align}
```
## doesn't stack

If you do multiple <kbd>alt</kbd> + <kbd>s</kbd> in succession (accidentally) you'll end up with impossible to parse, stacked environments. 
```
$$\begin{align}
\begin{align}\\
\\
\end{align}
\end{align}$$
```
(Erroneous nesting of equation structures)  
It should probably realise that it can execute alignify only once.

**Comment by GaurangTandon** [comment]
Thanks for the feedback! ^_^

> doesn't align right with multiple = in the same line

Is it that every single line having a `,` needs an ampersand before it? (I don't know, I am asking because whatever I write needs to be used by majority if not all the users) I have mostly only used `&` next to `=` signs or reaction arrows, or `&&` to create an empty column.

> align doesn't need dollars

"If MathJax is set up to parse environments processEnvironments: true, like it is on Chemistry.se", but what if it is not? ;) I need to cover every possible edge case as far as possible. And, personally, I did not find the `$$`s to be to intrusive. Is there any other reason why you would not want them to be there?

> doesn't recognise equality

My bad, will fix this in next release.

> doesn't stack

Aye, in the next release, pressing Alt-S again will unwrap the selection back to the original text.

**Comment by GaurangTandon** [comment]
With the [latest commit](https://github.com/GaurangTandon/MarkdownShortcutsForSE/commit/5fe2979e2e45ba2c94c38a650b95180ec2982bb1) (which I accidentally #4 instead of #5):

1. the default shortcut for `\pu` is `Alt-W` - to reduce finger travelling on the keyboard.
2. the code now smartly handles subscript and superscript tags. So, applying `ce` on `K<sub>2</sub>Cr<sub>2</sub>O<sub>7</sub>` gives `\ce{2Cr2O7}` directly. It also handles misplaced charges. So, `Ca<sup>+2</sup>` becomes `\ce{Ca^2+}` automatically. Similar stuff goes for applying  `pu`.
3. alignment is now applied to all of =, <, >, <=, >=, `\geq`, `\leq`.

Next release will fix unwrapping of `\frac` and `\align`.

**Comment by polyluxus** [comment]
Nice updates, reasonable key change... I was always looking for the shortcut in that area.

> Is it that every single line having a `,` needs an ampersand before it? (I don't know, I am asking because whatever I write needs to be used by majority if not all the users) I have mostly only used `&` next to `=` signs or reaction arrows, or `&&` to create an empty column.

If you would like to align multiple equations in one line correctly (automatic spacing), then they also will have to be separated by `&`. I realise, that this is something the script can't possibly handle, since it doesn't understand the context. However, we might come across times, where we have an equation like 
```
2\cos(\frac{\pi}{3}) = 2\cdot\frac{1}{2} = 1
2\sin(\frac{\pi}{3}) = 2\cdot\frac{\sqrt{3}}{2} = \sqrt{3}
```
and then the spacing is off:
```
$$\begin{align}
2\cos(\frac{\pi}{3}) &= 2\cdot\frac{1}{2} &= 1\\
2\sin(\frac{\pi}{3}) &= 2\cdot\frac{\sqrt{3}}{2} &= \sqrt{3}
\end{align}$$
```
In such cases it would be better (not perfect) to insert only the first `&`.

The example which I chose in the first post is (probably) one of the more common ones, i.e. stating an equation, then stating a condition. If that is handled in a sentence style, it should have a `,` or also `;` in it, which could be used as a delimiter. But I realise that is hard to do. Note though, that it should be `,&`, the separator comes after the delimiter. I don't think that can be handled always correctly by a script, I think you'd need an AI to sort that out.

I also realise, when I would code my example, I would code it on two lines:
```
\begin{align}
  \psi &= a_1\phi_1 + a_2\phi,&
  \text{where } a_1^2 + a_2^2 &= \sqrt2
\end{align}
```

There are other problems, but I'll file them separately.

---

> "If MathJax is set up to parse environments processEnvironments: true, like it is on Chemistry.se", but what if it is not? ;) I need to cover every possible edge case as far as possible. 

It is called **MarkdownShortcutsForSE** though. So why would you care about other sites. Also since the option is set by default, it would probably be with well intentions to assume the same here. After all, you can insert the maths with alt+e, like in other cases where you apply markup.

>And, personally, I did not find the `$$`s to be to intrusive. Is there any other reason why you would not want them to be there?
They are not intrusive, just superfluous. I don't like superfluous stuff. But I find the word 'superfluous' looks funny.

---

Regarding the second point of the updates:

If the script can handle `K<sub>2</sub>Cr<sub>2</sub>O<sub>7</sub>` to give `\ce{K2Cr2O7}`, can you make it change something like `C_{6}H_{12}O_{6}` to `\ce{C6H12O6}`, too?

**Comment by GaurangTandon** [comment]
> ...There are other problems, but I'll file them separately.

Yep, it was beginning to seem very hard already :/ Nvm though.

> So why would you care about other sites. Also since the option is set by default,

Oh, I just realized that the setting `processEnvironments: true` has been set on all SE sites. By covering every edge case, I had meant to say that I wished to cover every SE site (not other sites) as far as possible, because afaik, the setting `processEnvironments: true` can be changed by site mods. Anyway, I have removed the `$$` now. You're right that it's superfluous.

> If the script can handle `K<sub>2</sub>Cr<sub>2</sub>O<sub>7</sub>` to give `\ce{K2Cr2O7}`, can you make it change something like `C_{6}H_{12}O_{6}` to `\ce{C6H12O6}`, too?

Yep, why not! Good idea, I'll try that out, again for both `pu` and `ce`. I'll also be careful as you need underscores for denoting isotopes, [as seen here](https://mhchem.github.io/MathJax-mhchem/).

**Comment by GaurangTandon** [comment]
Final issue left in this thread is to support unwrapping of alignify and fracify.

**Comment by polyluxus** [comment]
Does it still inserts a `&` every time it finds equality?

**Comment by GaurangTandon** [comment]
Very sorry, the latest update had broken the equality sign insertion :( This is fixed now.

**Comment by polyluxus** [comment]
Let me briefly summarise: The script inserts `&` before each equality sign.  
I can live with that, after seeing that manual labour would be needed in any case if you are writing a one line aligned equation environment.

If you wanted to be on the very safe side, insert it only at the first one.

I just wanted to make sure I understand what the script is doing.

**Comment by GaurangTandon** [comment]
@polyluxus that's odd :/ I just tried out a reinstall of the userscript on the following text selection:

```
A=B
C=D=F
```

and it outputted:

```
\begin{align}
A&=B\\
C&=D=F
\end{align}
```

which I think is also what you expect to happen (insert the `&` for only the first equals sign in case there's more than one on the same line)

**Comment by polyluxus** [comment]
Okay, I don't know what happened there, I restarted Chrome and then it worked like you said. Thanks for clearing that up.

**Comment by GaurangTandon** [comment]
closed as all issues have been processed (#12 spun off into new issue)
